static/panes/diff.ts

```diff
@@ -0,0 +1,430 @@
+        };
+
+        this.onCompileResult(id + '_exec', compiler, compileResult);
+    }
+
+    override registerCallbacks() {
+        this.eventHub.on('executeResult', this.onExecuteResult, this);
+        this.eventHub.on('executor', this.onExecutor, this);
+        this.eventHub.on('executorClose', this.onExecutorClose, this);
+    }
+
+    requestResendResult(id: number | string) {
+        if (typeof id === 'string') {
+            const p = id.indexOf('_exec');
+            if (p !== -1) {
+                const execId = parseInt(id.substr(0, p));
+                this.eventHub.emit('resendExecution', execId);
+            }
+        } else {
+            this.eventHub.emit('resendCompilation', id);
+        }
+    }
+
+    override onCompiler(
+        id: number | string,
+        compiler: CompilerType | undefined,
+        options: unknown,
+        editorId: number,
+        treeId: number
+    ) {
+        if (!compiler) return;
+        options = options || '';
+        let name = compiler.name + ' ' + options;
+        // TODO: selectize doesn't play nicely with CSS tricks for truncation; this is the best I can do
+        // There's a plugin at: http://www.benbybenjacobs.com/blog/2014/04/09/no-wrap-plugin-for-selectize-dot-js
+        // but it doesn't look easy to integrate.
+        const maxLength = 30;
+        if (name.length > maxLength - 3) name = name.substr(0, maxLength - 3) + '...';
+        this.compilers[id] = {
+            id: id,
+            name: name,
+            options: options,
+            editorId: editorId,
+            treeId: treeId,
+            compiler: compiler,
+        };
+        if (this.lhs.id === undefined) {
+            this.lhs.compiler = this.compilers[id];
+            this.lhs.id = id;
+            this.onDiffSelect(id);
+        }
+        if (this.rhs.id === undefined) {
+            this.rhs.compiler = this.compilers[id];
+            this.rhs.id = id;
+            this.onDiffSelect(id);
+        }
+        this.updateCompilers();
+    }
+
+    onExecutor(id: number, compiler: CompilerType, options: unknown, editorId: number, treeId: number) {
+        this.onCompiler(id + '_exec', compiler, options, editorId, treeId);
+    }
+
+    override onCompilerClose(id: number | string) {
+        delete this.compilers[id];
+        this.updateCompilers();
+    }
+
+    onExecutorClose(id: number) {
+        this.onCompilerClose(id + '_exec');
+    }
+
+    override getDefaultPaneName() {
+        return 'Diff Viewer';
+    }
+
+    override getPaneTag() {
+        // this gets called during the super's constructor before lhs/rhs have been initialized
+        if ((this.lhs as any) !== undefined && (this.rhs as any) !== undefined) {
+            if (this.lhs.compiler && this.rhs.compiler) {
+                return `${this.lhs.compiler.name} vs ${this.rhs.compiler.name}`;
+            }
+        }
+        return '';
+    }
+
+    updateCompilersFor(selectize: TomSelect, id: number | string) {
```

Comment:
this.compilers is typed as `Record<string, ...>` doesn't that mean the number part of the union is wrong?

Reply:
The diff view differentiates between executor and compiler ids by appending `_exec` to editors, I'll update the type to be `Record<string | number, ...>`